shared-bindings/keypad/KeyMatrix.c

```diff
@@ -0,0 +1,263 @@
+//|         :ref:`lifetime-and-contextmanagers` for more info."""
+//|         ...
+//|
+STATIC mp_obj_t keypad_keymatrix___exit__(size_t n_args, const mp_obj_t *args) {
+    (void)n_args;
+    common_hal_keypad_keymatrix_deinit(args[0]);
+    return MP_ROM_NONE;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(keypad_keymatrix___exit___obj, 4, 4, keypad_keymatrix___exit__);
+
+STATIC void check_for_deinit(keypad_keymatrix_obj_t *self) {
+    if (common_hal_keypad_keymatrix_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+//|     num_keys: int
+//|     """The number of keys that are being scanned. (read-only)
+//|     """
+//|
+STATIC mp_obj_t keypad_keymatrix_get_num_keys(mp_obj_t self_in) {
+    keypad_keymatrix_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return MP_OBJ_NEW_SMALL_INT(common_hal_keypad_keymatrix_get_num_keys(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(keypad_keymatrix_get_num_keys_obj, keypad_keymatrix_get_num_keys);
+
+const mp_obj_property_t keypad_keymatrix_num_keys_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&keypad_keymatrix_get_num_keys_obj,
+              MP_ROM_NONE,
+              MP_ROM_NONE},
+};
+
+//|     def key_num(self, row: int, col: int) -> int:
+//|         """Return the key number for a given row and column.
+//|         The key number is calculated by ``row * len(col_pins) + col``.
+//|         """
+//|         ...
+//|
+STATIC mp_obj_t keypad_keymatrix_key_num(mp_obj_t self_in, mp_obj_t row_in, mp_obj_t col_in) {
+    keypad_keymatrix_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    check_for_deinit(self);
+
+    const mp_uint_t row = (mp_uint_t)mp_arg_validate_int_range(
+        mp_obj_get_int(row_in), 0, (mp_int_t)common_hal_keypad_keymatrix_get_num_rows(self), MP_QSTR_row);
+
+    const mp_int_t col = (mp_uint_t)mp_arg_validate_int_range(
+        mp_obj_get_int(col_in), 0, (mp_int_t)common_hal_keypad_keymatrix_get_num_cols(self), MP_QSTR_col);
+
+    return MP_OBJ_NEW_SMALL_INT(
+        (mp_int_t)common_hal_keypad_keymatrix_key_num(self, row, col));
+}
+MP_DEFINE_CONST_FUN_OBJ_3(keypad_keymatrix_key_num_obj, keypad_keymatrix_key_num);
+
+//|     def pressed(self, key_num: int) -> None:
+//|         """Return ``True`` if the given key is pressed. This is a debounced read
+//|         of the key state which bypasses the `events` `EventQueue`.
+//|         """
+//|         ...
+//|
+STATIC mp_obj_t keypad_keymatrix_pressed(mp_obj_t self_in, mp_obj_t key_num_in) {
+    keypad_keymatrix_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    check_for_deinit(self);
+
+    mp_uint_t key_num = mp_arg_validate_int_range(
+        mp_obj_get_int(key_num_in), 0, (mp_int_t)common_hal_keypad_keymatrix_get_num_keys(self), MP_QSTR_key_num);
+
+    return mp_obj_new_bool(common_hal_keypad_keymatrix_pressed(self, (mp_uint_t)key_num));
+}
+MP_DEFINE_CONST_FUN_OBJ_2(keypad_keymatrix_pressed_obj, keypad_keymatrix_pressed);
+
+//|     def store_states(self, states: _typing.WriteableBuffer) -> None:
```

Comment:
Same here, why not have a library do this?

Reply:
There are use cases for a fast path with no storage allocation.

Reply:
Leaving, but renamed to `get_states_into()`, for consistency.